presto-main/src/main/java/io/prestosql/operator/GroupedTopNRankAccumulator.java

```diff
@@ -0,0 +1,894 @@
+                : heapNodeBuffer.getRightChildHeapIndex(heapNodeIndex);
+    }
+
+    private void setChildIndex(long heapNodeIndex, HeapTraversal.Child child, long newChildIndex)
+    {
+        if (child == HeapTraversal.Child.LEFT) {
+            heapNodeBuffer.setLeftChildHeapIndex(heapNodeIndex, newChildIndex);
+        }
+        else {
+            heapNodeBuffer.setRightChildHeapIndex(heapNodeIndex, newChildIndex);
+        }
+    }
+
+    /**
+     * Pop the root node off the group ID's max heap.
+     *
+     * @param contextEvictionListener optional callback for the root node that gets popped off
+     */
+    private void heapPop(long groupId, @Nullable LongConsumer contextEvictionListener)
+    {
+        long heapRootNodeIndex = groupIdToHeapBuffer.getHeapRootNodeIndex(groupId);
+        checkArgument(heapRootNodeIndex != UNKNOWN_INDEX, "Group ID has an empty heap");
+
+        long lastHeapNodeIndex = heapDetachLastInsertionLeaf(groupId);
+        long lastPeerGroupIndex = heapNodeBuffer.getPeerGroupIndex(lastHeapNodeIndex);
+        long lastPeerGroupCount = heapNodeBuffer.getPeerGroupCount(lastHeapNodeIndex);
+
+        if (lastHeapNodeIndex == heapRootNodeIndex) {
+            // The root is the last node remaining
+            dropHeapNodePeerGroup(groupId, lastHeapNodeIndex, contextEvictionListener);
+        }
+        else {
+            // Pop the root and insert the last peer group back into the heap to ensure a balanced tree
+            heapPopAndInsert(groupId, lastPeerGroupIndex, lastPeerGroupCount, contextEvictionListener);
+        }
+
+        // peerGroupLookup entry will be updated by definition of inserting the last peer group into a new node
+        heapNodeBuffer.deallocate(lastHeapNodeIndex);
+    }
+
+    /**
+     * Detaches (but does not deallocate) the leaf in the bottom right-most position in the heap.
+     * <p>
+     * Given the fixed insertion order, the bottom right-most leaf will correspond to the last leaf node inserted into
+     * the balanced heap.
+     *
+     * @return leaf node index that was detached from the heap
+     */
+    private long heapDetachLastInsertionLeaf(long groupId)
+    {
+        long heapRootNodeIndex = groupIdToHeapBuffer.getHeapRootNodeIndex(groupId);
+        long heapSize = groupIdToHeapBuffer.getHeapSize(groupId);
+
+        long previousNodeIndex = UNKNOWN_INDEX;
+        HeapTraversal.Child childPosition = null;
+        long currentNodeIndex = heapRootNodeIndex;
+
+        heapTraversal.resetWithPathTo(heapSize);
+        while (!heapTraversal.isTarget()) {
+            previousNodeIndex = currentNodeIndex;
+            childPosition = heapTraversal.nextChild();
+            currentNodeIndex = getChildIndex(currentNodeIndex, childPosition);
+            verify(currentNodeIndex != UNKNOWN_INDEX, "Target node must exist");
+        }
+
+        // Detach the last insertion leaf node, but do not deallocate yet
+        if (previousNodeIndex == UNKNOWN_INDEX) {
```

Comment:
I think I still have a preference for this structure, because it makes it obvious what happens to the case when the previousNodeIndex value is not changed through certain paths through the while condition. Otherwise, readers have to make the mental connection of what happens when size == 1 vs previousNodeIndex value and heap traversal. Otherwise, afterwards, you still probably want to assert that previousNodeIndex != UNKNOWN_INDEX and give an explanation about the relationship with another condition above. To me this seems like less implict assumptions in code.

Reply:
Up to you. Generally, I prefer to keep special cases isolated (preferably early on) and have main/happy path unconditional. Extra `validate(previousNodeIndex != UNKNOWN_INDEX)` doesn't seem that bad IMO.